README.md
# clad [![Build Status](https://secure.travis-ci.org/plicease/App-clad.png)](http://travis-ci.org/plicease/App-clad)

Parallel SSH client

# SYNOPSIS

    clad [options] <cluster> <command>
    clad --list
    clad --help

# DESCRIPTION

Clad provides the ability to run the same command on several hosts at 
once.  The output is displayed unbuffered as the various hosts run the 
command.  The list of hosts is determined by reading a configuration file
which may also contain command aliases and environment settings.

# OPTIONS

- -n

    Dry run, just show the command that would be executed and each host.

- -a

    Do not colorize the host names in the output.

- -l user

    Specify a login name for all ssh connections.

- --verbose

    Print out a lot of debugging information which may be useful in debugging issues with clad.

- --serial

    Force clad to wait for the command to finish on each host before continuing to the next.  This
    will be slower, but may be easier to read the output.

- --config _name_

    Specify the name of an alternate configuration.  For example if you use `--config MyClad` then
    the configuration file `~/etc/MyClad.conf` will be used instead of `~/etc/Clad.conf`.

- --fat

    Send the server code with the payload and feed into Perl on the remote end.  This makes
    the total payload much larger, but it allows you to use clad with servers that do not
    have [App::clad](https://metacpan.org/pod/App::clad) installed.  The remote end must have Perl 5.6.1 or better in the `PATH`.

- --max _number_

    Limit the maximum number of simultaneous connections to `number`

- --file _filename_

    Copy files to the remote end as part of the payload.  May be specified multiple times.
    The names of the files are available as environment variables `FILE1`, `FILE2`, etc.  The
    files will automatically be removed on the remote end when the command completes.  An example
    usage for this would be to install rpm packages:

        % clad --file Database-Server-0.01-1.noarch.rp mycluster 'rpm -U $FILE1'

- --dir _directory_

    Recursively copy the directory to the remote end as part of the payload.  The name of the directory
    is available as an environment variable `DIR`.  The directory will automatically be removed on
    the remote end when the command completes.  For example if you are installing a directory full of
    rpm packages:

        % clad --dir ~/rpmbuild/RPMS/noarch mycluster 'rpm -U $DIR/*'

- --summary

    Do not print out standard output and standard input, just the exit values or signals returned from
    each host.

- --log-dir _dir_

    Specify a directory to write log files to.  Each host will have its own log file.

- --log

    Same as `--log-dir`, but the location is ~/clad/log.

- --purge

    Purge any logs that have collected under your home directory from using the `--log` option.

- --list

    List the clusters and aliases defined in your configuration.

- --help

    Print help and exit.

- --version

    Print the version and exit.

# CONFIGURATION

The configuration file is a [Clustericious::Config](https://metacpan.org/pod/Clustericious::Config) style configuration file.
See ["EXAMPLES"](#examples) for an example configuration. 
It contains these sections and configuration items:

## env

Environment hash to override environment variables on all hosts that run the command.

## cluster

Hash to define the clusters.  This is a hash of lists, where the keys are the cluster names
and the lists are the host names.  For example:

    ---
    cluster:
      mycluster:
        - host1
        - host2
      myothercluster:
        - host3
        - host4

You can use a single hostname not in the `cluster` section to specify a cluster
of one host, so long as it is a legal hostname understood by `ssh`.

## alias

Hash of aliases.  This is a useful place to specify common shortcuts.  The values
in this hash may be either strings or lists, allowing you to use the list or scalar
form of system.

## server\_command

[clad](https://metacpan.org/pod/clad) runs on both the client and the server.  This specifies the command used to
communicate with the client on the server end.  Unless you are testing [clad](https://metacpan.org/pod/clad) you
probably won't need to change this.

## fat

Include the server code as part of the payload.  This is useful for hosts that
do not already have [App::clad](https://metacpan.org/pod/App::clad) installed.  This is the same as the `--fat`
option above.

## fat\_server\_command

The command to execute on the server side when using the `--fat` command line
option or the `fat` configuration option.  The default is simply `perl`.

## ssh\_command

This is the `ssh` command to use on the client side.  It is `ssh` by default.

## ssh\_options

These are the `ssh` options used when opening a connection to the server.  The default
may change as needed.

## ssh\_extra

Extra ssh command line options to be added after `ssh_options`.  If you just want to
add a few options without replacing the existing set, this is the way to go.

## colors

A list of colors as understood by [Term::ANSIColor](https://metacpan.org/pod/Term::ANSIColor) which are used in alteration
for each host to help separate the output visually.

## fail\_color

Color to use if clad determined the remote call failed

- exit with non zero
- killed by signal
- failed to start (usually due to a bad command)

The default is `bold red`.

## err\_color

Color to use for output to standard error.  The default is `bold yellow`.

## script

A hash of inline scripts.  The keys are the script name and the values are the script
bodies.  For example, with

    ---
    script:
      dir_listing:
        #!/bin/bash
        for i in $( ls ); do
          echo item: $i
        done

You can get directory listing with

    % clad cluster dir_listing

# EXAMPLES

Here is an example configuration

    ---
    env:
      PATH: /home/starscream/perl5/bin:/usr/local/bin:/usr/bin:/bin
      PERL5LIB: /home/starscream/perl5/lib
    
    cluster:
      mailservers:
        - mail1
        - mail2
      webservers:
        - www1
        - www2
        - www3
      databases:
        - db1
        - db2
        - db3
        - db4
    
    alias:
      config_init: git clone git1:/cm/config-$CLUSTER.git ~/etc
      config_update: cd ~/etc && git pull
      config_destory: rm -rf ~/etc

## uptime

To find the uptime of the mailservers:

    % clad webservers uptime
    [mail1 out ]  21:27:04 up 4 days, 12:22,  0 users,  load average: 0.96, 1.01, 1.04
    [mail2 out ]  21:24:09 up 93 days, 12:52,  0 users,  load average: 1.25, 1.33, 1.29

To find the uptime of all servers in any cluster:

    % clad mailservers,webservers,databases
    [mail1 out ]  21:27:04 up 4 days, 12:22,  0 users,  load average: 0.96, 1.01, 1.04
    [mail2 out ]  21:24:09 up 93 days, 12:52,  0 users,  load average: 1.25, 1.33, 1.29
    [www1  out ]  21:24:37 up 93 days, 12:52,  0 users,  load average: 2.60, 2.34, 2.21
    [www2  out ]  21:23:06 up 93 days, 12:51,  0 users,  load average: 0.60, 0.50, 0.50
    [www3  out ]  21:24:05 up 93 days, 12:52,  0 users,  load average: 3.99, 3.62, 3.55
    [db1   out ]  21:24:53 up 93 days, 12:47,  0 users,  load average: 11.71, 12.15, 12.23
    [db2   out ]  21:26:07 up 93 days, 12:52,  0 users,  load average: 14.13, 13.91, 13.05
    [db3   out ]  21:29:06 up 93 days, 12:53,  0 users,  load average: 1.99, 1.59, 1.14
    [db4   out ]  21:24:55 up 93 days, 12:48,  0 users,  load average: 4.99, 4.83, 4.03

(note that the output in this example is displayed in order, though in practice it will usually be jumbled 

## log into hosts with different user

By default [clad](https://metacpan.org/pod/clad) will login to the remote servers with what ever user is default for `ssh`
(this is usually determined by the local user and / or the ssh configuration).  You can
use the `-l` option to specify a user name for all clusters in the command

    % clad -l foo mailservers,webservers,databases whoami
    [mail1 out ] foo
    [mail2 out ] foo
    [www1  out ] foo
    [www2  out ] foo
    [www3  out ] foo
    [db1   out ] foo
    [db2   out ] foo
    [db3   out ] foo
    [db4   out ] foo

or you can prefix individual clusters with a user name using the `@` sign.

    % clad foo@mailservers,bar@webservers,baz@database whoami
    [mail1 out ] foo
    [mail2 out ] foo
    [www1  out ] bar
    [www2  out ] bar
    [www3  out ] bar
    [db1   out ] baz
    [db2   out ] baz
    [db3   out ] baz
    [db4   out ] baz

## running Perl remotely

In the configuration above, we have specified `PATH` and `PERL5LIB` environment variables
to work with the modules build for [local::lib](https://metacpan.org/pod/local::lib) on each host (the actual configuration is
probably a little more complicated), so we can use modules that we have installed in
[local::lib](https://metacpan.org/pod/local::lib).

    % clad webservers -- perl -Mojo -E 'say g("mojolicio.us")->dom->at("title")->text'
    [www1 out ]
    [www1 out ]       Mojolicious - Perl real-time web framework
    [www1 out ]
    [www2 out ]
    [www2 out ]       Mojolicious - Perl real-time web framework
    [www2 out ]
    [www3 out ]
    [www3 out ]       Mojolicious - Perl real-time web framework
    [www3 out ]

## pulling remote configuration using git

[Clustericious](https://metacpan.org/pod/Clustericious) servers and client use configuration files that are usually stored in `~/etc`.
We usually manage these configurations on a cluster by cluster basis using git, and deploy
them using [clad](https://metacpan.org/pod/clad).

For example, to initialize the configuration directory using the &lt;config\_init> alias:

    ---
    alias:
      config_init: git clone git1:/cm/config-$CLUSTER.git ~/etc

and run:

    % clad webservers config_init

...we can update using the `config_update` alias:

    ---
    alias:
      config_update: cd ~/etc && git pull

and run:

    % clad webservers config_update

...and if the configuration becomes hosed, we can remove it and start over.
Since the master configuration is stored in git this may not be disaster.

    ---
    alias:
      config_destory: rm -rf ~/etc

and run:

    % clad webservers config_destroy

## using shell

[clad](https://metacpan.org/pod/clad) runs the command on the remote end using the same exact arguments as
you pass it on the client side.  That means that it uses either the single
argument or list version of `system` depending on input.  That means that
if you want to use shell logic, pipes or redirection, you need to use the
single argument version!  For example:

    % clad webservers cd ~/etc && git pull    # WRONG !
    % clad webservers 'cd ~/etc && git pull'  # RIGHT !

Sometimes if you don't want to worry about the escaping of meta characters
the list version will be more appropriate

    % clad webservers perl -E 'say "hi there"'

# ENVIRONMENT

## CLAD\_CLUSTER

This environment variable is set to the cluster name from the 
configuration file on each node that the command is run.

# INSTALL

You can override the default values for the `--fat`, `--server_command` and
`--fat_server_command` at install time using options to Build.PL.

    perl Build.PL --clad_fat \
                  --clad_server_command /usr/local/bin/perl \
                  --server_command /usr/local/bin/perl /usr/local/bin/clad --server

In this example, we specify fully qualified pathnames for Perl and [clad](https://metacpan.org/pod/clad), which
may be what you want in environments where the system Perl (usually installed in
`/usr/bin/perl`) comes before the Perl that you want to use.

# CAVEATS

[Clustericious::Admin](https://metacpan.org/pod/Clustericious::Admin) and [clad](https://metacpan.org/pod/clad) require an [AnyEvent](https://metacpan.org/pod/AnyEvent) event loop that allows
entering the event loop by calling `recv` on a condition variable.  This is not
supported by all [AnyEvent](https://metacpan.org/pod/AnyEvent) event loops and is discouraged by the [AnyEvent](https://metacpan.org/pod/AnyEvent)
documentation for CPAN modules, though most of the important event loops, such as
[EV](https://metacpan.org/pod/EV) and the pure perl implementation that comes with [AnyEvent](https://metacpan.org/pod/AnyEvent) DO support
this behavior.

# AUTHOR

Graham Ollis <plicease@cpan.org>

# COPYRIGHT AND LICENSE

This software is copyright (c) 2015-2019 by Graham Ollis.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.


Changes
Revision history for {{$dist->name}}

{{$NEXT}}
  - Undeprecated.

1.11      2018-03-11 14:21:46 -0400
  - Deprecated. See https://github.com/clustericious/Clustericious/issues/46

1.10      2017-07-22 22:34:12 -0400
  - Remove dependency on File::HomeDir

1.09      2017-03-14 10:59:11 -0400
  - Fix Test2 / Test::Exit order bug

1.08      2016-04-11 14:02:57 -0400
  - Migrated back to Module::Build.
  - --clad_server_command, --clad_fat_server_command and --clad_fat
    options to Build.PL can override the default values for the
    corresponding configuration items.

1.07      2016-01-31 12:08:35 -0500
  - As promised, older configuration keys aliases and clusters has been removed
  - Default output with no arguments is less verbose.
    To get the full help text, run clad --help or perldoc clad.

1.06      2015-12-16 15:13:04 -0500
  - Allow hostname as single host cluster.
  - Added --list option

1.05      2015-12-03 13:38:41 -0500
  - Added the --summary option
  - Added the --log, --log-dir and --purge options
  - Colorize failure and stderr (colors are configurable via fail_color and
    err_color configuration items)

1.04      2015-11-29 12:57:14 -0500
  - If STDIN is not a tty, then it will be read and
    passed onto the server.
  - Moved repository to clustericious org on github
    (updated meta appropriately)

1.03      2015-11-27 09:41:40 -0500
  - Added the --dir option
    (requires server 1.02)
  - Use Data::Dumper as fallback instead of JSON::PP.
    Thus in theory should work with a core Perl 5.6.x 
    server side.

1.02      2015-11-25 17:55:03 -0500
  - Internal performance tweaks

1.01      2015-11-25 21:42:53 +0000
  - Added the --file option
    (requires server 1.01)
  - Added script configuration item

1.00      2015-11-17 04:57:07 -0500
  - Minor documentation fixes, otherwise identical to
    the 0.99_05 release.

0.99_05   2015-11-16 20:52:07 -0500
  - Second release candidate for 1.00
  - Added --fat option (and fat config option) for
    hosts that don't already have App-clad installed
  - Relaxed requirement for server code (using --fat)
    to Perl 5.8 and JSON::PP.
  - Added --max option.

0.99_04   2015-11-16 06:57:26 -0500
  - Consider this to be a release candidate for 1.00
  - Stop argument parsing at first non-option,
    non recognized option, so arguments can be passed
    to remote end without -- (this fixes an issue
    mostly relevent since 0.99_01, which was a trial
    release anyway).

0.99_03   2015-11-16 04:27:10 -0500
  - Deprecate "clusters" in the configuration file
    in favor of "cluster".  "clusters" will be removed
    on or after January 31, 2016.
  - Restored the now deprecated (instead of removed)
    Clustericious::Admin interface.  Along with the
    other deprecations I plan on removing it on or
    after January 31, 2016.
  - MSWin32 is officially not supported.  It may be in
    a future revision.

0.99_02   2015-11-14 01:38:10 -0500
  - Added ssh_extra configuration useful for
    appending to ssh_options
  - Deprecate "aliases" in the configuration file
    in favor of "alias".  "aliases" will be removed
    on or after January 31, 2016.
  - Added some server side verification of payload.

0.99_01   2015-11-13 22:04:58 -0500
  - Complete rewrite from scratch.  List
    of (probably not exhaustive) changes
    to follow.
  - No support for macros.  These seem to have been
    documented in Clustericious::Admin, but
    not actually implemented.
  - Use of single argument or list version of
    system on the server side.  See EXAMPLES 
    for details.
  - Environment variables are more accurately passed
    to the server side when you are using shell
    logic operators (||, &&, etc) pipes ( | ).
  - Clustericious::Admin is no longer a public interface
    to clad.  It wasn't ever really inteded to be so,
    but it did include some documentation.  A public
    interface may be added here later.
  - Clustericious::Admin is an empty package included
    for dependency compatibility only.  The machinery
    for the app has been moved into App::clad.
  - App::clad and the clad executable is used on both
    the client AND the server.  This makes it more
    reliable, but does require that it be installed on
    both hosts, where before it only needed to be
    on the client.
  - This implementation uses AnyEvent::Open3::Simple and
    is object oriented to make the code easier to maintain.

0.25      2014-12-20 05:43:02 America/New_York
  - cpantesters do not report on older Perls

0.24      2014-08-19 16:17:41 America/New_York
  - migrate to Dist::Zilla
  - better documentation coverage
  - do not use ansi color if standard out is not a terminal

0.23      2013-08-19 07:36:09 America/New_York
  - increment version number this time

0.22      2013-08-19 07:34:14 America/New_York
  - minor documentation tweaks

0.21      2012-11-05 12:29:52 America/New_York
  - first public release



t/lib/Test/Clustericious/Config.pm
package Test::Clustericious::Config;

use strict;
use warnings;
use 5.010001;
use Test2::Plugin::FauxHomeDir;
use File::Glob qw( bsd_glob );
use YAML::XS qw( DumpFile );
use File::Path qw( mkpath );
use Clustericious::Config;
use Mojo::Loader;
use Test2::API qw( context );
use Test2::Mock;
use base qw( Exporter );

our @EXPORT = qw( create_config_ok create_directory_ok home_directory_ok create_config_helper_ok );
our @EXPORT_OK = @EXPORT;
our %EXPORT_TAGS = ( all => \@EXPORT );

my $config_dir;
my $mock;

sub _init
{
  $config_dir = bsd_glob('~/etc');
  mkdir $config_dir;
  $mock = Test2::Mock->new( class => 'Clustericious::Config::Util' );
  $mock->override(
    path => sub {
      ($config_dir)
    },
  );
}

BEGIN { _init() }

sub create_config_ok ($;$$)
{
  my($config_name, $config, $test_name) = @_;

  my $fn = "$config_name.conf";
  $fn =~ s/::/-/g;
  
  unless(defined $config)
  {
    my $caller = caller;
    Mojo::Loader::load_class($caller) unless $caller eq 'main';
    $config = Mojo::Loader::data_section($caller, "etc/$fn");
  }
  
  my @diag;
  my $config_filename;
  
  my $ctx = context();
  my $ok = 1;
  if(!defined $config)
  {
    $config = "---\n";
    push @diag, "unable to locate text for $config_name";
    $ok = 0;
    $test_name //= "create config for $config_name";
  }
  else
  {
    $config_filename = "$config_dir/$fn";
  
    eval {
      if(ref $config)
      {
        DumpFile($config_filename, $config);
      }
      else
      {
        open my $fh, '>', $config_filename;
        print $fh $config;
        close $fh;
      }
    };
    if(my $error = $@)
    {
      $ok = 0;
      push @diag, "exception: $error";
    }
  
    $test_name //= "create config for $config_name at $config_filename";
  
    # remove any cached copy if necessary
    Clustericious::Config::Util->uncache($config_name);
  }

  $ctx->ok($ok, $test_name);
  $ctx->diag($_) for @diag;  
  
  $ctx->release;
  
  return $config_filename;
}

sub create_directory_ok ($;$)
{
  my($path, $test_name) = @_;

  my $fullpath;
  my $ok;
  
  if(defined $path)
  {
    $fullpath = $path;
    $fullpath =~ s{^/}{};
    $fullpath = bsd_glob("~/$fullpath");
    mkpath $fullpath, 0, 0700;
  
    $test_name //= "create directory $fullpath";
    $ok = -d $fullpath;
  }
  else
  {
    $test_name //= "create directory [undef]";
    $ok = 0;
  }
  
  my $ctx = context();
  $ctx->ok($ok, $test_name);
  $ctx->release;
  return $fullpath;
}

sub home_directory_ok (;$)
{
  my($test_name) = @_;
  
  my $fullpath = bsd_glob('~');
  
  $test_name //= "home directory $fullpath";
  
  my $ctx = context();
  $ctx->ok(-d $fullpath, $test_name);
  $ctx->release;
  return $fullpath;
}

sub create_config_helper_ok ($$;$)
{
  my($helper_name, $helper_code, $test_name) = @_;
  
  $test_name //= "create config helper $helper_name";
  my $ok = 1;
   
  require Clustericious::Config::Helpers;
  do {
    no strict 'refs';
    *{"Clustericious::Config::Helpers::$helper_name"} = $helper_code;
  };
  push @Clustericious::Config::Helpers::EXPORT, $helper_name;
  
  my $ctx = context();
  $ctx->ok($ok, $test_name);
  $ctx->release;
  return $ok;
}

1;


lib/App/clad.pm
package App::clad;

use strict;
use warnings;
use 5.010;
use Getopt::Long 1.24 qw( GetOptionsFromArray :config pass_through);
use Pod::Usage qw( pod2usage );
use Clustericious::Config;
use Term::ANSIColor ();
use Sys::Hostname qw( hostname );
use YAML::XS qw( Dump );
use File::Basename qw( basename );
use File::Glob qw( bsd_glob );
use AE;
use Clustericious::Admin::RemoteHandler;
use Clustericious::Admin::Dump qw( perl_dump );
use File::chdir;
use Path::Class ();

# ABSTRACT: Parallel SSH client
# VERSION

=head1 SYNOPSIS

 % perldoc clad

=head1 DESCRIPTION

This module provides the implementation for the L<clad> command.  See 
the L<clad> command for the public interface.

=head1 SEE ALSO

=over 4

=item L<clad>

=back

=cut

sub _local_default ($$)
{
  eval { require App::clad::ConfigData }
    ? App::clad::ConfigData->config($_[0])
    : $_[1];
}

sub main
{
  my $clad = shift->new(@_);
  $clad->run;
}

# this hook is used for testing
# see t/args.t subtest 'color'
our $_stdout_is_terminal = sub { -t STDOUT };

sub _rc
{
  my $dir = bsd_glob('~/.clad');
  mkdir $dir unless $dir;
  $dir;
}

sub new
{
  my $class = shift;

  my $self = bless {
    dry_run    => 0,
    color      => $_stdout_is_terminal->(),
    server     => 0,
    verbose    => 0,
    serial     => 0,
    next_color => -1,
    ret        => 0,
    fat        => 0,
    max        => 0,
    count      => 0,
    summary    => 0,
    files      => [],
    purge      => 0,
    list       => 0,
  }, $class;
  
  my @argv = @_;
  
  my $config_name = 'Clad';
  
  GetOptionsFromArray(
    \@argv,
    'n'         => \$self->{dry_run},
    'a'         => sub { $self->{color} = 0 },
    'l=s'       => \$self->{user},
    'server'    => \$self->{server},
    'verbose'   => \$self->{verbose},
    'serial'    => \$self->{serial},
    'config=s'  => \$config_name,
    'fat'       => \$self->{fat},
    'max=s'     => \$self->{max},
    'file=s'    => $self->{files},
    'dir=s'     => \$self->{dir},
    'summary'   => \$self->{summary},
    'purge'     => \$self->{purge},
    'list'      => \$self->{list},

    'log'       => sub {
      $self->{log_dir} = Path::Class::Dir->new(
        _rc(),
        'log', 
        sprintf("%08x.%s", time, $$)
      );
    },

    'log-dir=s' => sub { $self->{log_dir} = Path::Class::Dir->new($_[1]) },

    'help|h'    => sub { pod2usage({ -verbose => 2}) },
    'version'   => sub {
      say STDERR 'App::clad version ', ($App::clad::VERSION // 'dev');
      exit 1;
    },
  ) || pod2usage(1);
  
  $self->log_dir->mkpath(0,0700) if $self->log_dir;

  $self->{config} = Clustericious::Config->new($config_name);

  return $self if $self->server;
  return $self if $self->purge;
  return $self if $self->list;
  
  # make sure there is at least one cluster specified
  # and that it doesn't look like a command line option
  unless(@argv)
  {
    pod2usage({
      -exitval  => 'NOEXIT',
      -message  => "No clusters specified", 
      -sections => [ qw( SYNOPSIS) ],
      -verbose  => 99,
    });
    $self->{list} = 1;
    $self->ret(1);
    return $self;
  }
  pod2usage({ -exitvalue => 1, -message => "Unknown option: $1" })
    if $argv[0] =~ /^--?(.*)$/;

  $self->{clusters} = [ split ',', shift @argv ];

  # make sure there is at least one command argument is specified
  # and that it doesn't look like a command line option
  pod2usage({ -exitval => 1, -message => "No commands specified" })
    unless @argv;
  pod2usage({ -exitvalue => 1, -message => "Unknown option: $1" })
    if $argv[0] =~ /^--?(.*)$/;
  
  $self->{command}  = [ @argv ];

  if(my $expanded = $self->alias->{$self->command->[0]})
  {
    if(ref $expanded)
    {
      splice @{ $self->command }, 0, 1, @$expanded;
    }
    else
    {
      $self->command->[0] = $expanded;
    }
  }
  
  if($self->config->script(default => {})->{$self->command->[0]})
  {
    my $name = shift @{ $self->command };
    unshift @{ $self->command }, '$SCRIPT1';
    my $content = $self->config->script(default => {})->{$name};
    $self->{script} = [ $name => $content ];
  }
  
  my $ok = 1;
  
  foreach my $cluster (map { "$_" } $self->clusters)
  {
    $cluster =~ s/^.*@//;
    unless($self->cluster_list->{$cluster})
    {
      $self->cluster_list->{$cluster} = [$cluster];
    }
  }
  
  foreach my $file ($self->files)
  {
    next if -r $file;
    say STDERR "unable to find $file";
    $ok = 0;
  }
  
  if(defined $self->dir && ! -d $self->dir)
  {
    say STDERR "unable to find @{[ $self->dir ]}";
    $ok = 0;
  }

  unless(-t STDIN)
  {
    $self->{stdin} = do { local $/; <STDIN> };
    delete $self->{stdin}
      unless defined $self->{stdin}
      &&     length $self->{stdin};
  }
  
  exit 2 unless $ok;
  
  $self;
}

sub config         { shift->{config}            }
sub dry_run        { shift->{dry_run}           }
sub color          { shift->{color}             }
sub clusters       { @{ shift->{clusters} }     }
sub command        { shift->{command}           }
sub user           { shift->{user}              }
sub server         { shift->{server}            }
sub verbose        { shift->{verbose}           }
sub serial         { shift->{serial}            }
sub max            { shift->{max}               }
sub files          { @{ shift->{files} }        }
sub dir            { shift->{dir}               }
sub script         { @{ shift->{script} // [] } }
sub stdin          { defined shift->{stdin}     }
sub summary        { shift->{summary}           }
sub log_dir        { shift->{log_dir}           }
sub purge          { shift->{purge}             }
sub list           { shift->{list}              }
sub fail_color     { shift->config->fail_color ( default => 'bold red'    ) }
sub err_color      { shift->config->err_color  ( default => 'bold yellow' ) }
sub ssh_command    { shift->config->ssh_command(    default => 'ssh' ) }
sub ssh_options    { shift->config->ssh_options(    default => [ -o => 'StrictHostKeyChecking=no', 
                                                                 -o => 'BatchMode=yes',
                                                                 -o => 'PasswordAuthentication=no',
                                                                 '-T', ] ) }
sub ssh_extra      { shift->config->ssh_extra(      default => [] ) }
sub fat            { my $self = shift; $self->{fat} || $self->config->fat( default => _local_default 'clad_fat', 0 ) }

sub server_command
{
  my($self) = @_;
  
  $self->fat
  ? $self->config->fat_server_command( default => _local_default 'clad_fat_server_command', 'perl' )
  : $self->config->server_command(     default => _local_default 'clad_server_command', 'clad --server' );
}

sub alias
{
  my($self) = @_;
  $self->config->alias( default => {} );
}

sub cluster_list
{
  my($self) = @_;
  $self->config->cluster( default => {} );
}

sub ret
{
  my($self, $new) = @_;
  $self->{ret} = $new if defined $new;
  $self->{ret};
}

sub host_length
{
  my($self) = @_;

  unless($self->{host_length})
  {
    my $length = 0;
  
    foreach my $cluster (map { "$_" } $self->clusters)
    {
      my $user = $cluster =~ s/^(.*)@// ? $1 : $self->user;
      foreach my $host (@{ $self->cluster_list->{$cluster} })
      {
        my $prefix = ($user ? "$user\@" : '') . $host;
        $length = length $prefix if length $prefix > $length;
      }
    }
    
    $self->{host_length} = $length;
  }
  
  $self->{host_length};
}

sub next_color
{
  my($self) = @_;
  my @colors = $self->config->colors( default => ['green','cyan'] );
  $colors[ ++$self->{next_color} ] // $colors[ $self->{next_color} = 0 ];
}

sub payload
{
  my($self, $clustername) = @_;
  
  my %env = $self->config->env( default => {} );
  $env{CLAD_CLUSTER} //= $clustername;

  my $payload = {
    env     => \%env,
    command => $self->command,
    verbose => $self->verbose,
    version => $App::clad::VERSION // 'dev',
  };
  
  if($self->files)
  {
    $payload->{require} = '1.01';
    
    foreach my $filename ($self->files)
    {
      my %h;
      open my $fh, '<', $filename;
      binmode $fh;
      $h{content} = do { local $/; <$fh> };
      close $fh;
      $h{name} = basename $filename;
      $h{mode} = sprintf "%o", (stat $filename)[2] & 0777;
      push @{ $payload->{files} }, \%h;
    }
  }
  
  if($self->script)
  {
    my($name, $content) = $self->script;
    $payload->{require} = '1.01';
    
    push @{ $payload->{files} }, {
      name    => $name,
      content => $content,
      mode    => '0700',
      env     => 'SCRIPT1',
    };
  }
  
  if($self->dir)
  {
    $payload->{require} = '1.02';
    
    $CWD = $self->dir;
    
    my $recurse;
    $recurse = sub {
      my($dir) = @_;
      foreach my $child ($dir->children(no_hidden => 1))
      {
        my $key = $child->relative->stringify;
        if($child->is_dir)
        {
          $payload->{dir}->{$key} = {
            is_dir => 1,
          };
          $recurse->($child);
        }
        else
        {
          $payload->{dir}->{$key} = {
            content => scalar $child->slurp(iomode => '<:bytes'),
          };
        }
        $payload->{dir}->{$key}->{mode} = sprintf '%o', $child->stat->mode & 0777;
      }
    };
    
    $recurse->(Path::Class::Dir->new);
  }

  if($self->stdin)
  {
    $payload->{require} = '1.04';
    
    # TODO:
    # In Perl 5.22 we could refalias this
    # and save some memory copies.
    $payload->{stdin} = $self->{stdin};
  }
  
  if($self->fat)
  {
    # Perl on the remote end may not have YAML
    # so we dump as Perl data structure
    # instead.
    $payload = perl_dump($payload);
    require Clustericious::Admin::Server;
    open my $fh, '<', $INC{'Clustericious/Admin/Server.pm'};
    my $code = do { local $/; <$fh> };
    close $fh;
    $code =~ s{\s*$}{"\n"}e;
    $payload = $code . $payload;
  }
  else
  {
    $payload = Dump($payload);
  }
  
  $payload;
}

sub run
{
  my($self) = @_;
  
  return $self->run_server if $self->server;
  return $self->run_purge  if $self->purge;
  return $self->run_list   if $self->list;
  
  my @done;
  my $max = $self->max;

  
  foreach my $cluster (map { "$_" } $self->clusters)
  {
    my $user = $cluster =~ s/^(.*)@// ? $1 : $self->user;

    my $payload = $self->payload($cluster);

    foreach my $host (@{ $self->cluster_list->{$cluster} })
    {
      my $prefix = ($user ? "$user\@" : '') . $host;
      if($self->dry_run)
      {
        say "$prefix % @{ $self->command }";
      }
      else
      {
        my $remote = Clustericious::Admin::RemoteHandler->new(
          prefix  => $prefix,
          clad    => $self,
          user    => $user,
          host    => $host,
          payload => $payload,
        );

        my $done = $remote->cv;
        
        $done->cb(sub {
          my $count = --$self->{count};
          $self->{cv}->send if $self->{cv};
        }) if $max;
        
        if($max)
        {
          my $count = ++$self->{count};
          if($count >= $max)
          {
            $self->{cv} = AE::cv;
            $self->{cv}->recv;
            delete $self->{cv};
          }
        }
        
        $self->serial ? $done->recv : push @done, $done;
      }
    }
  }
  
  $_->recv for @done;

  say "See @{[ $self->log_dir ]} for all logs" if $self->log_dir;
  
  $self->ret;
}

sub run_server
{
  require Clustericious::Admin::Server;
  Clustericious::Admin::Server->_server(*STDIN);
}

sub run_purge
{
  my $log_dir = Path::Class::Dir->new(
    _rc(),
    'log',
  );
  
  return unless -d $log_dir;
  
  foreach my $path ($log_dir->children)
  {
    if(-d $path)
    {
      say "PURGE DIR  $path";
      $path->rmtree(1, 1);
    }
    else
    {
      say "PURGE FILE $path";
      $path->remove;
    }
  }
}

sub run_list
{
  my($self) = @_;
  
  my @clusters = sort keys %{ $self->cluster_list };
  
  my $cluster = shift @clusters;
  if($cluster)
  {
    say "Clusters: $cluster";
    say "          $_" for @clusters;
  }
  else
  {
    say "Clusters: [none]";
  }

  my @alias = sort keys %{ $self->alias };
  my $alias = shift @alias;
  if($alias)
  {
    say "Aliases:  $alias";
    say "          $_" for @alias;
  }
  else
  {
    say "Aliases:  [none]";
  }

  $self->ret;
}

1;
